pkg/common/cns-lib/volume/util.go

```diff
@@ -161,3 +166,154 @@ func updateQueryResult(ctx context.Context, m *defaultManager, res *cnstypes.Cns
 	}
 	return res
 }
+
+// setupConnection connects to CNS and updates the VSphereUser to the session's user.
+func setupConnection(ctx context.Context, virtualCenter *cnsvsphere.VirtualCenter,
+	spec *cnstypes.CnsVolumeCreateSpec) error {
+	log := logger.GetLogger(ctx)
+	// Set up the VC connection.
+	err := virtualCenter.ConnectCns(ctx)
+	if err != nil {
+		log.Errorf("ConnectCns failed with err: %+v", err)
+		return err
+	}
+	// If the VSphereUser in the CreateSpec is different from session user,
+	// update the CreateSpec.
+	s, err := virtualCenter.Client.SessionManager.UserSession(ctx)
+	if err != nil {
+		log.Errorf("failed to get usersession with err: %v", err)
+		return err
+	}
+	if s.UserName != spec.Metadata.ContainerCluster.VSphereUser {
+		log.Debugf("Update VSphereUser from %s to %s", spec.Metadata.ContainerCluster.VSphereUser, s.UserName)
+		spec.Metadata.ContainerCluster.VSphereUser = s.UserName
+	}
+	return nil
+}
+
+// getPendingCreateVolumeTaskFromMap returns the CreateVolume task for a volume stored in the volumeTaskMap.
+func getPendingCreateVolumeTaskFromMap(ctx context.Context, volNameFromInputSpec string) *object.Task {
+	var task *object.Task
+	log := logger.GetLogger(ctx)
+	taskDetailsInMap, ok := volumeTaskMap[volNameFromInputSpec]
+	if ok {
+		task = taskDetailsInMap.task
+		log.Infof("CreateVolume task still pending for Volume: %q, with taskInfo: %+v",
+			volNameFromInputSpec, task)
+	}
+	return task
+}
+
+// invokeCNSCreateVolume truncates the input volume name and invokes a CreateVolume operation for that volume on CNS.
+func invokeCNSCreateVolume(ctx context.Context, virtualCenter *cnsvsphere.VirtualCenter,
+	spec *cnstypes.CnsVolumeCreateSpec) (*object.Task, error) {
+	var cnsCreateSpecList []cnstypes.CnsVolumeCreateSpec
+	log := logger.GetLogger(ctx)
+	// Truncate the volume name to make sure the name is within 80 characters before calling CNS.
+	if len(spec.Name) > maxLengthOfVolumeNameInCNS {
+		volNameAfterTruncate := spec.Name[0 : maxLengthOfVolumeNameInCNS-1]
+		log.Infof("Create Volume with name %s is too long, truncate it to %s", spec.Name, volNameAfterTruncate)
+		spec.Name = volNameAfterTruncate
+		log.Debugf("CNS Create Volume is called with %v", spew.Sdump(*spec))
+	}
+	cnsCreateSpecList = append(cnsCreateSpecList, *spec)
+	task, err := virtualCenter.CnsClient.CreateVolume(ctx, cnsCreateSpecList)
+	if err != nil {
+		log.Errorf("CNS CreateVolume failed from vCenter %q with err: %v", virtualCenter.Config.Host, err)
```

Comment:
Error messages should start with lower case. Can you check this in this PR?

Reply:
Actually from https://github.com/golang/go/wiki/CodeReviewComments#error-strings

```
... Error strings should not be capitalized (unless beginning with proper nouns or acronyms) ...
```
So since this message starts with `CNS` i've kept it as is.